src/trel_dnssd/trel_dnssd.cpp

```diff
@@ -0,0 +1,542 @@
+{
+    mTrelNetif = aTrelNetif;
+
+    if (IsInitialized())
+    {
+        otbrLogDebug("Initialized on netif \"%s\"", aTrelNetif.c_str());
+        CheckTrelNetifReady();
+    }
+    else
+    {
+        otbrLogDebug("Not initialized");
+    }
+}
+
+void TrelDnssd::StartBrowse(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Start browsing %s service ...", kTrelServiceName);
+
+    assert(mSubscriberId == 0);
+    mSubscriberId = mPublisher.AddSubscriptionCallbacks(
+        [this](const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo) {
+            OnTrelServiceDiscovered(aType, aInstanceInfo);
+        },
+
+        nullptr);
+
+    if (IsReady())
+    {
+        mPublisher.SubscribeService(kTrelServiceName, "");
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::StopBrowse(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Stop browsing %s service.", kTrelServiceName);
+    assert(mSubscriberId > 0);
+
+    mPublisher.RemoveSubscriptionCallbacks(mSubscriberId);
+    mSubscriberId = 0;
+
+    if (IsReady())
+    {
+        mPublisher.UnsubscribeService(kTrelServiceName, "");
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::RegisterService(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength)
+{
+    assert(aPort > 0);
+    assert(aTxtData != nullptr);
+
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Register %s service: port=%u, TXT=%d bytes", kTrelServiceName, aPort, aTxtLength);
+    otbrDump(OTBR_LOG_DEBUG, OTBR_LOG_TAG, "TXT", aTxtData, aTxtLength);
+
+    if (mRegisterInfo.IsValid() && IsReady())
+    {
+        DoRemoveTrelService();
+    }
+
+    mRegisterInfo.Assign(aPort, aTxtData, aTxtLength);
+
+    if (IsReady())
+    {
+        DoRegisterTrelService();
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::RemoveService(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Remove %s service.", kTrelServiceName);
+    assert(mRegisterInfo.IsValid());
+
+    if (IsReady())
+    {
+        DoRemoveTrelService();
+    }
+
+    mRegisterInfo.Clear();
+
+exit:
+    return;
+}
+
+void TrelDnssd::OnMdnsPublisherReady(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("mDNS Publisher is Ready");
+    mMdnsPublisherReady = true;
+    RemoveAllPeers();
+
+    if (mRegisterInfo.IsRegistered())
+    {
+        mRegisterInfo.mInstanceName = "";
+    }
+
+    OnBecomeReady();
+
+exit:
+    return;
+}
+
+void TrelDnssd::OnTrelServiceDiscovered(const std::string &                            aType,
+                                        const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo)
+{
+    VerifyOrExit(StringUtils::EqualCaseInsensitive(aType, kTrelServiceName));
+    VerifyOrExit(aInstanceInfo.mNetifIndex == mTrelNetifIndex);
+
+    if (aInstanceInfo.mRemoved)
+    {
+        RemoveTrelService(aInstanceInfo.mName);
+    }
+    else
+    {
+        OnTrelServiceAdded(aInstanceInfo);
+    }
```

Comment:
How is `OnTrelServiceDiscovered` called when something in an existing (previously discovered) service get changed? For example, if the original device that registered the service, changes the TXT records.
- Do we get a `aInstanceInfo.mRemoved` first, followed by another call, adding it again with new info.
- Or do we get a call with new info (I guess this may be the case).

Reply:
> Do we get a aInstanceInfo.mRemoved first, followed by another call, adding it again with new info.

Not necesssary have `Remove`. I think it's common for a service to be discovered and updated multiple times. And TREL core can also handle that.

Reply:
On the other hand, `OnTrelServiceAdded` will always call `RemoveTrelService(instanceName);` before actually adding. 
So, TREL core will always get a remove and then another add.